File: node/src/mcp-server.ts
#!/usr/bin/env node
/**
 * Standalone MCP server exposing HAI SDK operations as tools.
 *
 * Usage:
 *   haisdk-mcp                           # stdio transport
 *   npx tsx src/mcp-server.ts            # dev mode
 *   HAI_URL=https://hai.ai haisdk-mcp    # override API endpoint
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';
import { fileURLToPath } from 'node:url';
import { HaiClient } from './client.js';
import { generateVerifyLink } from './verify.js';

export async function getClient(args: Record<string, unknown>): Promise<HaiClient> {
  const options: Record<string, unknown> = {};
  if (typeof args.hai_url === 'string' && args.hai_url) {
    options.url = args.hai_url;
  }
  if (typeof args.config_path === 'string' && args.config_path) {
    options.configPath = args.config_path;
  }
  return HaiClient.create(options as any);
}

function toJSON(obj: unknown): string {
  return JSON.stringify(obj);
}

export const TOOLS = [
  // ----- Identity tools -----
  {
    name: 'hai_hello',
    description: 'Run authenticated hello handshake with HAI',
    inputSchema: {
      type: 'object' as const,
      properties: {
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_check_username',
    description: 'Check if a hai.ai username is available',
    inputSchema: {
      type: 'object' as const,
      properties: {
        username: { type: 'string', description: 'Username to check' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['username'],
    },
  },
  {
    name: 'hai_claim_username',
    description: 'Claim a hai.ai username for an agent',
    inputSchema: {
      type: 'object' as const,
      properties: {
        agent_id: { type: 'string', description: 'Agent UUID' },
        username: { type: 'string', description: 'Username to claim' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['agent_id', 'username'],
    },
  },
  {
    name: 'hai_register_agent',
    description: 'Register the local JACS agent with HAI',
    inputSchema: {
      type: 'object' as const,
      properties: {
        owner_email: { type: 'string', description: 'Owner email address' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_agent_status',
    description: "Get the current agent's verification status",
    inputSchema: {
      type: 'object' as const,
      properties: {
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_verify_agent',
    description: "Verify another agent's JACS document",
    inputSchema: {
      type: 'object' as const,
      properties: {
        agent_document: { type: 'string', description: 'JACS document (JSON string)' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['agent_document'],
    },
  },
  {
    name: 'hai_generate_verify_link',
    description: 'Generate a HAI verify link from a signed JACS document',
    inputSchema: {
      type: 'object' as const,
      properties: {
        document: { type: 'string', description: 'Signed JACS document JSON string' },
        base_url: { type: 'string', description: 'Verifier base URL override' },
        hosted: { type: 'boolean', description: 'Use hosted verify URL mode' },
      },
      required: ['document'],
    },
  },
  // ----- Email tools -----
  {
    name: 'hai_send_email',
    description: "Send an email from the agent's @hai.ai address",
    inputSchema: {
      type: 'object' as const,
      properties: {
        to: { type: 'string', description: 'Recipient email address' },
        subject: { type: 'string', description: 'Email subject line' },
        body: { type: 'string', description: 'Plain text email body' },
        in_reply_to: { type: 'string', description: 'Message-ID to reply to (for threading)' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['to', 'subject', 'body'],
    },
  },
  {
    name: 'hai_list_messages',
    description: "List email messages in the agent's inbox/outbox",
    inputSchema: {
      type: 'object' as const,
      properties: {
        limit: { type: 'integer', description: 'Max messages to return (default 20)' },
        offset: { type: 'integer', description: 'Pagination offset' },
        direction: { type: 'string', description: "Filter: 'inbound' or 'outbound'" },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_get_message',
    description: 'Get a single email message by ID',
    inputSchema: {
      type: 'object' as const,
      properties: {
        message_id: { type: 'string', description: 'Message UUID' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['message_id'],
    },
  },
  {
    name: 'hai_delete_message',
    description: 'Delete an email message',
    inputSchema: {
      type: 'object' as const,
      properties: {
        message_id: { type: 'string', description: 'Message UUID' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['message_id'],
    },
  },
  {
    name: 'hai_mark_read',
    description: 'Mark an email message as read',
    inputSchema: {
      type: 'object' as const,
      properties: {
        message_id: { type: 'string', description: 'Message UUID' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['message_id'],
    },
  },
  {
    name: 'hai_mark_unread',
    description: 'Mark an email message as unread',
    inputSchema: {
      type: 'object' as const,
      properties: {
        message_id: { type: 'string', description: 'Message UUID' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['message_id'],
    },
  },
  {
    name: 'hai_search_messages',
    description: 'Search email messages by query, sender, recipient, or date range',
    inputSchema: {
      type: 'object' as const,
      properties: {
        q: { type: 'string', description: 'Search query text' },
        direction: { type: 'string', description: "Filter: 'inbound' or 'outbound'" },
        from_address: { type: 'string', description: 'Filter by sender address' },
        to_address: { type: 'string', description: 'Filter by recipient address' },
        since: { type: 'string', description: 'Filter: messages after this ISO date' },
        until: { type: 'string', description: 'Filter: messages before this ISO date' },
        limit: { type: 'integer', description: 'Max results (default 20)' },
        offset: { type: 'integer', description: 'Pagination offset' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_get_unread_count',
    description: 'Get the count of unread email messages',
    inputSchema: {
      type: 'object' as const,
      properties: {
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_get_email_status',
    description: 'Get email account status including usage limits and daily stats',
    inputSchema: {
      type: 'object' as const,
      properties: {
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
    },
  },
  {
    name: 'hai_reply_email',
    description: 'Reply to an email message (fetches original, sends reply with threading)',
    inputSchema: {
      type: 'object' as const,
      properties: {
        message_id: { type: 'string', description: 'ID of the message to reply to' },
        body: { type: 'string', description: 'Reply body text' },
        subject_override: { type: 'string', description: 'Override the Re: subject line' },
        config_path: { type: 'string', description: 'Path to jacs.config.json' },
        hai_url: { type: 'string', description: 'HAI API URL override' },
      },
      required: ['message_id', 'body'],
    },
  },
];

function textResult(text: string) {
  return { content: [{ type: 'text' as const, text }] };
}

function errorResult(message: string) {
  return { content: [{ type: 'text' as const, text: message }], isError: true as const };
}

export async function handleToolCall(
  name: string,
  args: Record<string, unknown>,
): Promise<{ content: { type: 'text'; text: string }[]; isError?: true }> {
  try {
    const client = await getClient(args);

    switch (name) {
      // Identity
      case 'hai_hello': {
        const result = await client.hello();
        return textResult(toJSON(result));
      }
      case 'hai_check_username': {
        const result = await client.checkUsername(args.username as string);
        return textResult(toJSON(result));
      }
      case 'hai_claim_username': {
        const result = await client.claimUsername(args.agent_id as string, args.username as string);
        return textResult(toJSON(result));
      }
      case 'hai_register_agent': {
        const opts: Record<string, unknown> = {};
        if (args.owner_email) opts.ownerEmail = args.owner_email;
        const result = await client.register(opts as any);
        return textResult(toJSON(result));
      }
      case 'hai_agent_status': {
        const result = await client.verify();
        return textResult(toJSON(result));
      }
      case 'hai_verify_agent': {
        const result = await client.verifyAgent(args.agent_document as string);
        return textResult(toJSON(result));
      }
      case 'hai_generate_verify_link': {
        const result = generateVerifyLink(
          args.document as string,
          typeof args.base_url === 'string' ? args.base_url : undefined,
          Boolean(args.hosted),
        );
        return textResult(toJSON({ verify_url: result }));
      }

      // Email
      case 'hai_send_email': {
        const result = await client.sendEmail({
          to: args.to as string,
          subject: args.subject as string,
          body: args.body as string,
          inReplyTo: (args.in_reply_to as string) || undefined,
        });
        return textResult(toJSON(result));
      }
      case 'hai_list_messages': {
        const result = await client.listMessages({
          limit: args.limit as number | undefined,
          offset: args.offset as number | undefined,
          direction: args.direction as 'inbound' | 'outbound' | undefined,
        });
        return textResult(toJSON(result));
      }
      case 'hai_get_message': {
        const result = await client.getMessage(args.message_id as string);
        return textResult(toJSON(result));
      }
      case 'hai_delete_message': {
        await client.deleteMessage(args.message_id as string);
        return textResult(JSON.stringify({ deleted: true, message_id: args.message_id }));
      }
      case 'hai_mark_read': {
        await client.markRead(args.message_id as string);
        return textResult(JSON.stringify({ message_id: args.message_id, is_read: true }));
      }
      case 'hai_mark_unread': {
        await client.markUnread(args.message_id as string);
        return textResult(JSON.stringify({ message_id: args.message_id, is_read: false }));
      }
      case 'hai_search_messages': {
        const result = await client.searchMessages({
          query: (args.q as string) || '',
          limit: args.limit as number | undefined,
          offset: args.offset as number | undefined,
          direction: args.direction as 'inbound' | 'outbound' | undefined,
          fromAddress: args.from_address as string | undefined,
          toAddress: args.to_address as string | undefined,
        });
        return textResult(toJSON(result));
      }
      case 'hai_get_unread_count': {
        const count = await client.getUnreadCount();
        return textResult(JSON.stringify({ count }));
      }
      case 'hai_get_email_status': {
        const result = await client.getEmailStatus();
        return textResult(toJSON(result));
      }
      case 'hai_reply_email': {
        const result = await client.reply(
          args.message_id as string,
          args.body as string,
          (args.subject_override as string) || undefined,
        );
        return textResult(toJSON(result));
      }

      default:
        return errorResult(`unknown tool: ${name}`);
    }
  } catch (err) {
    return errorResult(err instanceof Error ? err.message : String(err));
  }
}

export async function main() {
  const server = new Server(
    { name: 'hai-sdk', version: '0.1.0' },
    { capabilities: { tools: {} } },
  );

  server.setRequestHandler(ListToolsRequestSchema, async () => ({
    tools: TOOLS,
  }));

  server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;
    return handleToolCall(name, (args ?? {}) as Record<string, unknown>);
  });

  const transport = new StdioServerTransport();
  await server.connect(transport);
}

const currentFile = fileURLToPath(import.meta.url);
const invokedFile = process.argv[1];

if (invokedFile && currentFile === invokedFile) {
  main().catch((err) => {
    process.stderr.write(`server error: ${err}\n`);
    process.exit(1);
  });
}

File: python/src/haisdk/mcp_server.py
"""Standalone MCP server exposing HAI SDK operations as tools.

Usage:
    haisdk-mcp                           # stdio transport
    python -m haisdk.mcp_server          # alternative
    HAI_URL=https://hai.ai haisdk-mcp    # override API endpoint
"""

from __future__ import annotations

import json
import os
from dataclasses import asdict
from typing import Optional

from mcp.server.fastmcp import FastMCP

server = FastMCP("hai-sdk")

_DEFAULT_URL = os.environ.get("HAI_URL", "https://hai.ai")


def _url(hai_url: str | None) -> str:
    return hai_url or _DEFAULT_URL


def _to_json(obj: object) -> str:
    if hasattr(obj, "__dataclass_fields__"):
        return json.dumps(asdict(obj), default=str)  # type: ignore[arg-type]
    if isinstance(obj, (dict, list)):
        return json.dumps(obj, default=str)
    return str(obj)


# ---------------------------------------------------------------------------
# Identity tools
# ---------------------------------------------------------------------------


@server.tool()
async def hai_hello(
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Run authenticated hello handshake with HAI."""
    from jacs.hai.client import hello_world

    result = hello_world(_url(hai_url))
    return _to_json(result)


@server.tool()
async def hai_register_agent(
    owner_email: str | None = None,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Register the local JACS agent with HAI."""
    from jacs.hai.client import register

    result = register(_url(hai_url), owner_email=owner_email)
    return _to_json(result)


@server.tool()
async def hai_agent_status(
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Get the current agent's verification status."""
    from jacs.hai.client import status

    result = status(_url(hai_url))
    return _to_json(result)


@server.tool()
async def hai_check_username(
    username: str,
    hai_url: str | None = None,
) -> str:
    """Check if a hai.ai username is available."""
    from jacs.hai.client import check_username

    result = check_username(_url(hai_url), username)
    return _to_json(result)


@server.tool()
async def hai_claim_username(
    agent_id: str,
    username: str,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Claim a hai.ai username for an agent."""
    from jacs.hai.client import claim_username

    result = claim_username(_url(hai_url), agent_id, username)
    return _to_json(result)


@server.tool()
async def hai_verify_agent(
    agent_document: str,
    hai_url: str | None = None,
) -> str:
    """Verify another agent's JACS document."""
    from jacs.hai.client import verify_agent

    result = verify_agent(agent_document, hai_url=_url(hai_url))
    return _to_json(result)


@server.tool()
async def hai_generate_verify_link(
    document: str,
    base_url: str | None = None,
    hosted: bool = False,
) -> str:
    """Generate a HAI verify link from a signed JACS document."""
    from jacs.hai.client import generate_verify_link

    url = generate_verify_link(document, base_url=base_url or "https://hai.ai", hosted=hosted)
    return json.dumps({"verify_url": url})


# ---------------------------------------------------------------------------
# Email tools
# ---------------------------------------------------------------------------


@server.tool()
async def hai_send_email(
    to: str,
    subject: str,
    body: str,
    in_reply_to: str | None = None,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Send an email from the agent's @hai.ai address."""
    from jacs.hai.client import send_email

    result = send_email(_url(hai_url), to=to, subject=subject, body=body, in_reply_to=in_reply_to)
    return _to_json(result)


@server.tool()
async def hai_list_messages(
    limit: int = 20,
    offset: int = 0,
    direction: str | None = None,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """List email messages in the agent's inbox/outbox."""
    from jacs.hai.client import list_messages

    result = list_messages(_url(hai_url), limit=limit, offset=offset, direction=direction)
    return _to_json(result)


@server.tool()
async def hai_get_message(
    message_id: str,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Get a single email message by ID."""
    from jacs.hai.client import get_message

    result = get_message(_url(hai_url), message_id)
    return _to_json(result)


@server.tool()
async def hai_delete_message(
    message_id: str,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Delete an email message."""
    from jacs.hai.client import delete_message

    delete_message(_url(hai_url), message_id)
    return json.dumps({"deleted": True, "message_id": message_id})


@server.tool()
async def hai_mark_read(
    message_id: str,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Mark an email message as read."""
    from jacs.hai.client import mark_read

    mark_read(_url(hai_url), message_id)
    return json.dumps({"message_id": message_id, "is_read": True})


@server.tool()
async def hai_mark_unread(
    message_id: str,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Mark an email message as unread."""
    from jacs.hai.client import mark_unread

    mark_unread(_url(hai_url), message_id)
    return json.dumps({"message_id": message_id, "is_read": False})


@server.tool()
async def hai_search_messages(
    q: str | None = None,
    direction: str | None = None,
    from_address: str | None = None,
    to_address: str | None = None,
    since: str | None = None,
    until: str | None = None,
    limit: int = 20,
    offset: int = 0,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Search email messages by query, sender, recipient, or date range."""
    from jacs.hai.client import search_messages

    result = search_messages(
        _url(hai_url),
        q=q,
        direction=direction,
        from_address=from_address,
        to_address=to_address,
        since=since,
        until=until,
        limit=limit,
        offset=offset,
    )
    return _to_json(result)


@server.tool()
async def hai_get_unread_count(
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Get the count of unread email messages."""
    from jacs.hai.client import get_unread_count

    count = get_unread_count(_url(hai_url))
    return json.dumps({"count": count})


@server.tool()
async def hai_get_email_status(
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Get email account status including usage limits and daily stats."""
    from jacs.hai.client import get_email_status

    result = get_email_status(_url(hai_url))
    return _to_json(result)


@server.tool()
async def hai_reply_email(
    message_id: str,
    body: str,
    subject_override: str | None = None,
    config_path: str | None = None,
    hai_url: str | None = None,
) -> str:
    """Reply to an email message (fetches original, sends reply with threading)."""
    from jacs.hai.client import reply

    result = reply(_url(hai_url), message_id, body, subject=subject_override)
    return _to_json(result)


def main() -> None:
    """Entry point for haisdk-mcp CLI command."""
    server.run(transport="stdio")


if __name__ == "__main__":
    main()
